### ISSUE TITLE
scp server returns non-zero exit code to remote client

### ISSUE THREAD

**Issue by girtsf** [ronf/asyncssh#308]
When running the server in `scp` mode, even successful transfers result in non-zero exit code for the client (commandline `scp`).

I can reproduce this with the following server code:
```
import asyncio, asyncssh, sys


class MySSHServer(asyncssh.SSHServer):
    def connection_made(self, conn):
        self._conn = conn

    def begin_auth(self, username):
        try:
            self._conn.set_authorized_keys(username)
        except IOError:
            pass

        return True


async def start_server():
    await asyncssh.create_server(
        MySSHServer,
        '',
        8022,
        server_host_keys=['ssh_host_key'],
        sftp_factory=True,
        allow_scp=True)


loop = asyncio.get_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))

loop.run_forever()
```

And by running `scp -P 8022 some_file user@localhost:some_file` in the terminal.

After quick fiddling, I got it to exit cleanly with the following patch, but I wasn't sure if there are any other implications that I've missed.

```
--- scp.py.bak	2020-09-01 14:01:34.008416937 -0700
+++ scp.py	2020-09-01 14:01:51.849024770 -0700
@@ -308,8 +308,9 @@
     async def close(self):
         """Close an SCP session"""

         self.logger.info('Stopping remote SCP')
+        self._writer.channel.exit(0)
         self._writer.close()
         await self._writer.channel.wait_closed()
```

**Comment by ronf** [comment]
Thanks very much for catching and reporting this, and identifying the source of the problem.

I've been able to reproduce this here. In fact, I already had a test program which triggered it, but I never looked at the return code and the command-line scp client doesn't actually report an error of any kind. However, it apparently silently exits with an exit status of 1 when the SSH connection is closed without an exit status being sent.

Unfortunately, the fix is going to be a bit more complicated than what you proposed. Your change does actually work for the case where AsyncSSH is acting as an scp server. However, that code path is also used in the case where the channel is an SSHClientChannel attached to an outbound SSH connection, and there is no exit() method available in that case. I'm still looking into the best way of handling that. I'll follow up here once I have more information.

**Comment by ronf** [comment]
Ok - I think the following works:

```patch
diff --git a/asyncssh/scp.py b/asyncssh/scp.py
index 96429e2..7ab13c5 100644
--- a/asyncssh/scp.py
+++ b/asyncssh/scp.py
@@ -167,10 +167,11 @@ class _SCPArgParser(argparse.ArgumentParser):
 class _SCPHandler:
     """SCP handler"""
 
-    def __init__(self, reader, writer, error_handler=None):
+    def __init__(self, reader, writer, error_handler=None, server=False):
         self._reader = reader
         self._writer = writer
         self._error_handler = error_handler
+        self._server = server
 
         self._logger = reader.logger.get_child('sftp')
 
@@ -309,7 +310,12 @@ class _SCPHandler:
         """Close an SCP session"""
 
         self.logger.info('Stopping remote SCP')
-        self._writer.close()
+
+        if self._server:
+            self._writer.channel.exit(0)
+        else:
+            self._writer.close()
+
         await self._writer.channel.wait_closed()
 
 
@@ -318,8 +324,8 @@ class _SCPSource(_SCPHandler):
 
     def __init__(self, fs, reader, writer, preserve, recurse,
                  block_size=SFTP_BLOCK_SIZE, progress_handler=None,
-                 error_handler=None):
-        super().__init__(reader, writer, error_handler)
+                 error_handler=None, server=False):
+        super().__init__(reader, writer, error_handler, server)
 
         self._fs = fs
         self._preserve = preserve
@@ -451,8 +457,8 @@ class _SCPSink(_SCPHandler):
 
     def __init__(self, fs, reader, writer, must_be_dir, preserve, recurse,
                  block_size=SFTP_BLOCK_SIZE, progress_handler=None,
-                 error_handler=None):
-        super().__init__(reader, writer, error_handler)
+                 error_handler=None, server=False):
+        super().__init__(reader, writer, error_handler, server)
 
         self._fs = fs
         self._must_be_dir = must_be_dir
@@ -934,9 +940,9 @@ def run_scp_server(sftp_server, command, stdin, stdout, stderr):
 
     if args.source:
         handler = _SCPSource(fs, stdin, stdout, args.preserve, args.recurse,
-                             error_handler=False)
+                             error_handler=False, server=True)
     else:
         handler = _SCPSink(fs, stdin, stdout, args.must_be_dir, args.preserve,
-                           args.recurse, error_handler=False)
+                           args.recurse, error_handler=False, server=True)
 
     return run_handler()
```

When launched as a server, it makes the exit(0) call on the SSHServerChannel. When started as a client, it just does a plain close() on the SSHClientChannel as it did before.

This change is now in the "develop" branch as commit 2cdaf6d and will be rolled into the next release.

Thanks again!

**Comment by girtsf** [comment]
Thanks for the quick fix and thanks for asyncssh in general, it's great!

**Comment by ronf** [comment]
My pleasure... This fix is now included in AsyncSSH 2.4.1. Thanks for your help on it!
